syntax = "proto3";

package transaction_verification;

service TransactionVerificationService {
    rpc InitializeOrder(InitializationRequest) returns (ResponseData);
    rpc VerifyCreditCardNumber(RequestData) returns (ResponseData);
    rpc VerifyCreditCardExpiryDate(RequestData) returns (ResponseData);
    rpc VerifyOrderItems (RequestData) returns (ResponseData);
    rpc VerifyTransaction(RequestData) returns (ResponseData);
}

message InitializationRequest {
    string orderId = 1;
    repeated TransactionItem items = 2;
    string userName = 3;
    string userContact = 4;
    string discountCode = 5;
    BillingAddressInfo billingAddress = 6;
    CreditCardInfo creditCard = 7;
}

message RequestData {
    string orderId = 1;
    repeated int32 vectorClock = 2;
}

message ResponseData {
    bool isSuccess = 1;
}

message TransactionItem {
    string name = 1;
    int32 quantity = 2;
}

message BillingAddressInfo {
    string street = 1;
    string city = 2;
    string state = 3;
    string zip = 4;
    string country = 5;
}

message CreditCardInfo {
    string number = 1;
    string expirationDate = 2;
    string cvv = 3;
}
